homeassistant/components/tesla/__init__.py

```diff
@@ -34,69 +40,154 @@
     extra=vol.ALLOW_EXTRA,
 )
 
-NOTIFICATION_ID = "tesla_integration_notification"
-NOTIFICATION_TITLE = "Tesla integration setup"
+
+@callback
+def _async_save_tokens(hass, config_entry, access_token, refresh_token):
+    hass.config_entries.async_update_entry(
+        config_entry,
+        data={
+            **config_entry.data,
+            CONF_ACCESS_TOKEN: access_token,
+            CONF_TOKEN: refresh_token,
+        },
+    )
 
 
 async def async_setup(hass, base_config):
     """Set up of Tesla component."""
-    config = base_config.get(DOMAIN)
 
+    def _update_entry(email, scan_interval=300, data=None, options=None):
+        data = data or {}
+        options = options or {CONF_SCAN_INTERVAL: scan_interval}
+        for entry in hass.config_entries.async_entries(DOMAIN):
+            if email == entry.title:
+                hass.config_entries.async_update_entry(
+                    entry, data=data, options=options
+                )
+
+    config = base_config.get(DOMAIN)
+    if not config:
+        return True
     email = config.get(CONF_USERNAME)
     password = config.get(CONF_PASSWORD)
-    update_interval = config.get(CONF_SCAN_INTERVAL)
-    if hass.data.get(DOMAIN) is None:
+    scan_interval = config.get(CONF_SCAN_INTERVAL, 300)
+    if email in configured_instances(hass):
+        for entry in hass.config_entries.async_entries(DOMAIN):
+            if email == entry.title:
+                try:
+                    websession = aiohttp_client.async_get_clientsession(hass)
+                    controller = teslaAPI(
+                        websession, email=email, password=password, update_interval=300
+                    )
+                    (refresh_token, access_token) = await controller.connect(
+                        test_login=True
+                    )
+                    hass.config_entries.async_update_entry(
+                        entry,
+                        data={
+                            CONF_ACCESS_TOKEN: access_token,
+                            CONF_TOKEN: refresh_token,
+                        },
+                        options={CONF_SCAN_INTERVAL: scan_interval},
+                    )
+                except TeslaException as ex:
+                    _LOGGER.warning(
+                        "Configuration.yaml credentials cannot communicate with Tesla API: %s",
+                        ex.message,
+                    )
+                break
+    else:
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": SOURCE_IMPORT},
+                data={CONF_USERNAME: email, CONF_PASSWORD: password},
+            )
+        )
+        async_call_later(hass, 15, _update_entry(email, scan_interval))
```

Comment:
`_update_entry` will be called immediately here so the call later won't do what is needed.

Reply:
I'm not sure of the best way to actually do this without a lambda. Any thoughts?

Reply:
hmm good point Martin, maybe we should allow it ?

Reply:
Yeah, as we talked about before in another PR, I would like the option to use the config flow to update both config entry data and options, for an existing config entry.

It would also be good to allow creating options data for new entries in the config flow from an import. Some config yaml options will map to options data and not config entry data, and there might be validation in the options flow that we want to avoid duplicating.

These are the possible tasks that I'm suggesting:
1. Update existing config entry data from config yaml import.
2. Update existing config entry options data from config yaml import.
3. Create new config entry with options data from config yaml import.

Reply:
Right. 1 and 2 are already possible inside `async_setup` via `hass.config_entries.async_update_entry(entry, data=X, options=Y)`. It's just 3 that's missing.

Reply:
The existing solution for 1 and 2 is what's being used here and which is what is awkward.

Reply:
Well we shouldn't call later, why can't we update it right away ?

Reply:
I don't know.

Reply:
@balloob In my test system there was a race condition where the entry wasn't created yet so we couldn't find it to update. The `_update_entry` function searches for a matching config entry and uses `async_update_entry` to update the option.  Once we used `call later` the option was updating properly.

Reply:
We should trigger this update in `async_setup_entry`. That's the hook we can use to know that the config entry has been created.

Reply:
@MartinHjelmare How would we pass the `scan_interval` through? It's a bit convoluted to pass an import only variable through the `config_flow`. Or are you suggesting it get saved to a `hass.data` dictionary?

Reply:
Yeah, that's the best option we have at the moment, to save the config in `hass.data` and update the config entry in `async_setup_entry` from the config data in `hass.data`.

Reply:
Ok done.